[DO NOT REVIEW] engine: Implement Engine, Iterator interfaces for Pebble

itsbilal opened the pull request:
This change implements the interfaces defined in
`pkg/storage/engine/engine.go` for Pebble, in particular `Engine`,
`Iterator`, and related ones. Also ports the MVCC Scanner to Go
for use in `iter.MVCC{Get,Scan}`.

Other related changes elsewhere in the codebase to let an incomplete
Engine implementer (i.e. without WithSSTables, and without a compactor)
run otherwise.

Also adds a parameter in the `--store` command line flag to specify
what engine to use. `--store=engine=pebble,path=node1` starts a Pebble
based store.

Fixes #39674.

TODO:
- [ ] Refactor tests to run for both RocksDB and Pebble Engines, instead
of one-or-the-other
- [ ] Add more test coverage, especially around batches
- [ ] Function signature comments
- [ ] Comments comments comments

Release note: None

Commit:
[WIP] engine: Implement Engine, Iterator interfaces for Pebble

This change implements the interfaces defined in
`pkg/storage/engine/engine.go` for Pebble, in particular `Engine`,
`Iterator`, and related ones. Also ports the MVCC Scanner to Go
for use in `iter.MVCC{Get,Scan}`.

Other related changes elsewhere in the codebase to let an incomplete
Engine implementer (i.e. without WithSSTables, and without a compactor)
run otherwise.

Also adds a parameter in the `--store` command line flag to specify
what engine to use. `--store=engine=pebble,path=node1` starts a Pebble
based store.

Fixes #39674.

TODO:
- [ ] Refactor tests to run for both RocksDB and Pebble Engines, instead
of one-or-the-other
- [ ] Add more test coverage, especially around batches
- [ ] Function signature comments
- [ ] Comments comments comments

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/41137)
<!-- Reviewable:end -->

itsbilal commented:
Putting this out early to make higher-level design more transparent. Don't bother code-reviewing it just yet, there's a lot still to clean up (especially in the tests).

I'll soon split the mega commit into three, one with just the interface implementations, one with the command-line stuff + any changes to let this Engine work, and one with all test-related changes.

petermattis reviewed (commented):
Nice progress!

This looks in line with what I was expecting. I've added a few random comments below during an initial skim. One comment that I think would be good to address before too long is to split up `pebble.go`. The monolithic `rocksdb.go` is a historic artifact. I can imagine `pebble_iterator.go`, `pebble_batch.go`, and `pebble_mvcc_scanner.go`. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41137#-:-LpjPO_dBAXF6FvXJXg2:b-8ur6yq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal)

___
*[pkg/server/config.go, line 489 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41137#-LpjC6J2A4F6V97sswXD:-LpjC6J2A4F6V97sswXE:b-2ltslt) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fd1f8d6adb2648ff734d38d45ae439c737390f5/pkg/server/config.go#L489)):*
> ```Go
> 			if spec.Engine == base.EngineTypePebble {
> 				pebbleOpts := &pebble.Options{
> 					Cache: pebbleCache.New(128 << 20),
> ```

`pebbleCache.New` is awkward. Perhaps `pebble/cache` should be moved to `pebble/internal/cache` and we expose cache creation via `pebble.NewCache`.

Note that for RocksDB, we share a single cache instance across multiple RocksDB instances. We should do the same for pebble.
___
*[pkg/storage/engine/pebble.go, line 478 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41137#-LpjD_xs1S8wyzfYKVO7:-LpjD_xs1S8wyzfYKVO8:bqnymwq) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fd1f8d6adb2648ff734d38d45ae439c737390f5/pkg/storage/engine/pebble.go#L478)):*
> ```Go
> 
> func (p *Pebble) OpenFile(filename string) (DBFile, error) {
> 	fs := vfs.Default
> ```

These `*File` ops should all use `p.opts.FS`.
___
*[pkg/storage/engine/pebble.go, line 537 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41137#-LpjDfCIEf-l8FKkpmM2:-LpjDfCIEf-l8FKkpmM3:b15bze2) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fd1f8d6adb2648ff734d38d45ae439c737390f5/pkg/storage/engine/pebble.go#L537)):*
> ```Go
> 
> func (p *Pebble) CreateCheckpoint(_ string) error {
> 	// No-op: Pebble does not implement checkpoints.
> ```

Let's make this a `TODO`. Seems like adding support for checkpoints may be helpful for debugging.
___
*[pkg/storage/engine/pebble.go, line 768 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41137#-LpjE-BvA1jSt_5c2sOE:-LpjE-BvA1jSt_5c2sOF:bqrhsxk) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fd1f8d6adb2648ff734d38d45ae439c737390f5/pkg/storage/engine/pebble.go#L768)):*
> ```Go
> 
> func (p *pebbleIterator) SeekReverse(key MVCCKey) {
> 	p.Seek(key)
> ```

I'm surprised this isn't using `SeekLT`, though I always get confused about the semantic mismatch of `SeekReverse` and `SeekLT`. I believe `SeekReverse` is effectively `SeekLessThanOrEqual`. 
___
*[pkg/storage/engine/pebble.go, line 811 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41137#-LpjEwCh5FyyPvwdXpYg:-LpjEwCh5FyyPvwdXpYh:b-jldwl1) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fd1f8d6adb2648ff734d38d45ae439c737390f5/pkg/storage/engine/pebble.go#L811)):*
> ```Go
> 
> func (p *pebbleIterator) ValueProto(msg protoutil.Message) error {
> 	value := p.Value()
> ```

I believe you can use `UnsafeValue` here. Proto unmarshaling will copy out data it needs.
___
*[pkg/storage/engine/pebble.go, line 914 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41137#-LpjFDPSCkmflGKzCJp8:-LpjFDPSCkmflGKzCJp9:b-5uvkrq) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fd1f8d6adb2648ff734d38d45ae439c737390f5/pkg/storage/engine/pebble.go#L914)):*
> ```Go
> 
> // Go port of mvccScanner in libroach/mvcc.h .
> type pebbleMvccScanner struct {
> ```

This is hefty enough that I think it deserves to be in its own file: `pebble_mvcc_scan.go` or `pebble_mvcc_scanner.go`.
___
*[pkg/storage/engine/pebble.go, line 1595 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41137#-LpjFlvyEs6vFo52ajla:-LpjFlvyEs6vFo52ajlb:bqyqg8l) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fd1f8d6adb2648ff734d38d45ae439c737390f5/pkg/storage/engine/pebble.go#L1595)):*
> ```Go
> 	closed       bool
> 	isDistinct   bool
> 	writeOnly    bool
> ```

I'm not sure if you need to track this separately. `pebble.Batch` will complain if you try to read from a non-indexed batch.
___
*[pkg/storage/engine/pebble.go, line 1596 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41137#-LpjFtnJDM4CdBtpesVG:-LpjFtnJDM4CdBtpesVH:b11wr4z) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fd1f8d6adb2648ff734d38d45ae439c737390f5/pkg/storage/engine/pebble.go#L1596)):*
> ```Go
> 	isDistinct   bool
> 	writeOnly    bool
> 	distinctOpen bool
> ```

Distinct batches are a performance optimization to avoid Cgo crossings. I wonder if they are necessary in the pebble implementation. I'm not suggesting we remove them at this point, but when we get to performance testing we should see if they provide value. 
___
*[pkg/storage/engine/pebble.go, line 1733 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41137#-LpjGL4H80TZ8RMSE8PF:-LpjGL4H80TZ8RMSE8PG:bt04de7) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fd1f8d6adb2648ff734d38d45ae439c737390f5/pkg/storage/engine/pebble.go#L1733)):*
> ```Go
> 	}
> 
> 	// Use the cached iterator.
> ```

The iterator caching was another optimization to avoid Cgo crossings. It might not be necessary with pebble. Again, this is good for now, but we should take a look if it is necessary with pebble. Let's add a TODO here so we don't forget.
___
*[pkg/storage/engine/pebble.go, line 1768 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41137#-LpjGa3Y7mFjcijVbu3t:-LpjGa3Y7mFjcijVbu3u:b-6jusmr) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fd1f8d6adb2648ff734d38d45ae439c737390f5/pkg/storage/engine/pebble.go#L1768)):*
> ```Go
> 	p.buf = EncodeKeyToBuf(p.buf[:0], key)
> 	err := p.batch.Delete(p.buf, nil)
> 	if err == pebble.ErrNotFound {
> ```

Deletes are blind in both RocksDB and Pebble. That is, they don't check to see if the key is present. You shouldn't see `ErrNotFound` here or for `SingleDelete`.


<!-- Sent from Reviewable.io -->

itsbilal commented:
Thanks for the early review! I've addressed your comments, done almost all cleanup I could think of, and addressed most of the TODOs above. Since I expect this PR to snowball for a while, I'll reset reviewable history by closing/reopening a PR on the same branch.
